### ISSUE TITLE
comment lines cause crash. AttributeError in yapflib/comment_splicer.py, line 105

### ISSUE THREAD

**Issue by kkheller** [google/yapf#144]
First try at running yapf crashed on every py file I tried.  I made my own (dirty-sloppy) patch for comment_splicer.py, and since then things are running great.

The AttributeError crash seems so fundamental (any/every comment line causes this crash?) that i can only conclude that there *must* be something weird in my own environment, otherwise the whole world would have this problem, and yapf would be unusable.

Describing my patch seems like it could be potentially useful in any case.

The crash was like so:

      File "/Users/kkheller/Downloads/yapf/yapf/yapflib/comment_splicer.py", line 105, in _VisitNodeRec
        for leaf_in_parent in stmt_parent.leaves():
    AttributeError: 'Node' object has no attribute 'leaves'

My dirty-sloppy patch is "roughly" the following:

    diff --git a/yapf/yapflib/comment_splicer.py b/yapf/yapflib/comment_splicer.py
    index 12a8eff..ee752cc 100644
    --- a/yapf/yapflib/comment_splicer.py
    +++ b/yapf/yapflib/comment_splicer.py
    @@ -102,6 +102,8 @@ def SpliceComments(tree):
                 # and is part of some expression.
                 stmt_parent = _FindStmtParent(child)
     
    +            try:
    +
                 for leaf_in_parent in stmt_parent.leaves():
                   if leaf_in_parent.type == token.NEWLINE:
                     continue
    @@ -132,6 +134,16 @@ def SpliceComments(tree):
                     pytree_utils.InsertNodesBefore(comments, child)
                     break
     
    +            except:
    +                  # This comment stands on its own line, and it has to be inserted
    +                  # into the appropriate parent. We'll have to find a suitable
    +                  # parent to insert into. See comments above
    +                  # _STANDALONE_LINE_NODES for more details.
    +                  node_with_line_parent = _FindNodeWithStandaloneLineParent(child)
    +                  pytree_utils.InsertNodesBefore(_CreateCommentsFromPrefix(
    +                      comment_prefix, comment_lineno,
    +                      standalone=True), node_with_line_parent)
    +
             prev_leaf[0] = child
     
       _VisitNodeRec(tree)

That is not literally the whole patch. The part that I have omitted for brevity is all the required indenting of all the code that is now inside of the added "try".

Describing what I patched using prose: at line 105, when stmt_parent.leaves() causes the error about the Node object not having any "leaves" attribute, what i tried to do was make that case (the error case) behave exactly like the case that was labeled with the comment "This comment stands on its own line, and it has to be inserted into the appropriate parent".

In other words: the preexisting code was already written to decide that a comment "stands on its own line" when execution enters the branch "elif id(leaf_in_parent) == id(child)".  My sloppy-dirty try/except (with a blatant copy-paste of the innards of that elif block) was intended to *also* treat the comment the same way whenever it turned out that the Node object did not have any "leaves" attribute (the AttributeError case).  

That was really just a "lucky guess" of how I needed to patch things.  It immediately enabled me to run yapf successfully on dozens of files, and I am pleased with everything yapf did to those files, so i haven't reflected further on whether my patch really is the proper fix or not. (and of *course* the copy-paste part of it is not ok)

My information:

I get the crash when using the v0.1.8 release, and when using eafa1fec6b from master.

I am running yapf without installing it.  I run it like so:

PYTHONPATH=$PYTHONPATH:~/Downloads/yapf/yapf python ~/Downloads/yapf  -i test.py 

also like so:

PYTHONPATH=~/Downloads/yapf/yapf python ~/Downloads/yapf  -i test.py

My python is:
Python 2.7 (r27:82508, Jul  3 2010, 21:12:11) 
[GCC 4.0.1 (Apple Inc. build 5493)] on darwin

Here is a "test.py" that i can achieve the crash with:

    import os

    def start_session():
        # need to retrieve some kind of 'config object'
        pass

**Comment by kkheller** [comment]
i mistyped my command line earlier. this doesn't change anything else in my report, just the command. should be:

PYTHONPATH=~/Downloads/yapf python ~/Downloads/yapf/yapf -r -i .

**Comment by gwelymernans** [comment]
This is very odd. The situation you're describing literally shouldn't happen. :-( Could you do me a favor and add these two lines to the top of `SpliceComments`?

```
  from yapf.yapflib import pytree_visitor                                                         
  print('node: {}'.format(pytree_visitor.DumpPyTree(tree)))                                       
```

The output should tell us what `lib2to3` is producing for your `test.py` example.

**Comment by kkheller** [comment]
Hi, Bill. Thank you for taking an interest in this. I always think it is fun to learn something bizarre and unexpected from totally edge-case bugs. I hoped someone else would be curious (someone with more python experience than i have).

Here is the output now:


    kkheller$ PYTHONPATH=~/Downloads/yapf python ~/Downloads/yapf/yapf -i test.py
    file_input [3 children] [child_indent="None"]
      simple_stmt [2 children] [child_indent="None"]
        import_name [2 children] [child_indent="None"]
          NAME(Leaf(1, u'import')) [lineno=1, column=0, prefix=u'']
          NAME(Leaf(1, u'os')) [lineno=1, column=7, prefix=u' ']
        NEWLINE(Leaf(4, u'\n')) [lineno=1, column=9, prefix='']
      funcdef [5 children] [child_indent="None"]
        NAME(Leaf(1, u'def')) [lineno=3, column=0, prefix=u'\n']
        NAME(Leaf(1, u'start_session')) [lineno=3, column=4, prefix=u' ']
        parameters [2 children] [child_indent="None"]
          LPAR(Leaf(7, u'(')) [lineno=3, column=17, prefix='']
          RPAR(Leaf(8, u')')) [lineno=3, column=18, prefix='']
        COLON(Leaf(11, u':')) [lineno=3, column=19, prefix='']
        suite [4 children] [child_indent="None"]
          NEWLINE(Leaf(4, u'\n')) [lineno=3, column=20, prefix='']
          INDENT(Leaf(5, u'    ')) [lineno=5, column=0, prefix=u"    # need to retrieve some kind of 'config object'\n"]
          simple_stmt [2 children] [child_indent="None"]
            NAME(Leaf(1, u'pass')) [lineno=5, column=4, prefix='']
            NEWLINE(Leaf(4, u'\n')) [lineno=5, column=8, prefix='']
          DEDENT(Leaf(6, '')) [lineno=6, column=0, prefix='']
      ENDMARKER(Leaf(0, '')) [lineno=6, column=0, prefix='']
    node: None
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 162, in _run_module_as_main
        "__main__", fname, loader, pkg_name)
      File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
        exec code in run_globals
      File "/Users/kkheller/Downloads/yapf/yapf/__main__.py", line 16, in <module>
        yapf.run_main()
      File "/Users/kkheller/Downloads/yapf/yapf/__init__.py", line 223, in run_main
        sys.exit(main(sys.argv))
      File "/Users/kkheller/Downloads/yapf/yapf/__init__.py", line 153, in main
        verify=args.verify)
      File "/Users/kkheller/Downloads/yapf/yapf/__init__.py", line 187, in FormatFiles
        print_diff=print_diff, verify=verify)
      File "/Users/kkheller/Downloads/yapf/yapf/yapflib/yapf_api.py", line 79, in FormatFile
        verify=verify), encoding
      File "/Users/kkheller/Downloads/yapf/yapf/yapflib/yapf_api.py", line 107, in FormatCode
        comment_splicer.SpliceComments(tree)
      File "/Users/kkheller/Downloads/yapf/yapf/yapflib/comment_splicer.py", line 140, in SpliceComments
        _VisitNodeRec(tree)
      File "/Users/kkheller/Downloads/yapf/yapf/yapflib/comment_splicer.py", line 54, in _VisitNodeRec
        _VisitNodeRec(child)
      File "/Users/kkheller/Downloads/yapf/yapf/yapflib/comment_splicer.py", line 54, in _VisitNodeRec
        _VisitNodeRec(child)
      File "/Users/kkheller/Downloads/yapf/yapf/yapflib/comment_splicer.py", line 108, in _VisitNodeRec
        for leaf_in_parent in stmt_parent.leaves():
    AttributeError: 'Node' object has no attribute 'leaves'

I hope this ends up being valuable in some way.

**Comment by kkheller** [comment]
I noticed something about that DumpPyTree output.  

Most of the string literals have the 'u' in front of them.  Such as:

    u'import'
    u'os'
    u'\n'

However, many of the empty strings do not have the u prefix.

The very first occurrence of an empty string, however, does have a u prefix.  So it cannot be said that every empty string lacks the u.

    NAME(Leaf(1, u'import')) [lineno=1, column=0, prefix=u'']
    
I am not sure if what I noticed is meaningful, or perhaps just a superficial weakness about how output is formatted.

**Comment by gwelymernans** [comment]
I'm not sure if that's the problem. I'm seeing the same thing when I output the pytree.

The pytree that I dump is identical to yours. I'm not sure what to do. If you don't mind, would you be up for helping to track this down? If you could sprinkle some `print` statements around `comment_slicer.py` to see what exactly it's failing on, I might be able to help interpret them. :-)

**Comment by kkheller** [comment]
i would *love* to sprinkle print statements! :)

if you want to make a branch (or gist? or something? ...i might not know the best thing, but i'll go with what you choose), then i could just check out (or patch, or merge) files that have the "print"s inserted to your liking, and run that.  

also, i will try to reproduce this later today on 2 other mac computers near my desk. the computer where i encounter this bug/issue is my oldest, and it has accumulated the most cruft over the years (things installed and uninstalled, paths changed, locales and region&currency settings messed with -- the works!).

**Comment by kkheller** [comment]
Here are my additional reproducibility results:

I repeated my steps on two other mac computers. The newest (and least "crufty") computer is running Mac OS X 10.9.4.

I was able to reproduce the bug on these two computers, too -- which brings the total count of affected computers to 3.

However...

While the reproducibility made me feel less inclined to place blame upon the cruftiness of my first computer, I now have reason to place blame on python flavor "[GCC 4.0.1 (Apple Inc. build 5493)] on darwin".

All 3 affected computers were running the python version "[GCC 4.0.1 (Apple Inc. build 5493)] on darwin".

On one of the computers, I installed python-2.7.8-macosx10.6.dmg from python.org, and the issue disappeared!  I would imagine that for 99% of people who would run into this issue, installing python-2.7.8-macosx10.6.dmg would be a totally reasonable solution.  So it's ok to close this if you wish.

After installing 2.7.8 the full version/flavor looks like this:
Python 2.7.8 (v2.7.8:ee879c0ffa11, Jun 29 2014, 21:07:35) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin

On the other hand, i still have 2 computers where i can sprinkle print statements :)  And I am happy to still do that.
